Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp

```diff
@@ -214,71 +341,109 @@ static RefPtr<SharedBuffer> parseResponseMessage(const String& message)
     return SharedBuffer::create(message.characters8() + offset, message.sizeInBytes() - offset);
 }
 
-Ref<CDMInstanceOpenCDM::Session> CDMInstanceOpenCDM::Session::create(const media::OpenCdm& source, Ref<WebCore::SharedBuffer>&& initData)
+Ref<CDMInstanceOpenCDM::Session> CDMInstanceOpenCDM::Session::create(OpenCDMAccessor& source, const String& keySystem, const char* type, Ref<WebCore::SharedBuffer>&& initData, LicenseType licenseType, Ref<WebCore::SharedBuffer>&& customData)
 {
-    return adoptRef(*new Session(source, WTFMove(initData)));
+    return adoptRef(*new Session(source, keySystem, type, WTFMove(initData), licenseType, WTFMove(customData)));
 }
 
-CDMInstanceOpenCDM::Session::Session(const media::OpenCdm& source, Ref<WebCore::SharedBuffer>&& initData)
-    : m_session(source)
-    , m_initData(WTFMove(initData))
+CDMInstanceOpenCDM::Session::Session(OpenCDMAccessor& source, const String& keySystem, const char* mimeType, Ref<WebCore::SharedBuffer>&& initData, LicenseType licenseType, Ref<WebCore::SharedBuffer>&& customData)
+    : m_initData(WTFMove(initData))
 {
-    // FIXME: This can be a source of overflows and other big errors. This should be fixed in the framework and then fixed here.
-    uint8_t temporaryURL[1024];
-    uint16_t temporaryURLLength = sizeof(temporaryURL);
-
-    std::string message;
-    m_session.GetKeyMessage(message, temporaryURL, temporaryURLLength);
+    auto& callbacksNotifier = OCDMSessionCallbacksNotifier::singleton();
 
-    if (message.empty() || !temporaryURLLength)
+    OpenCDMSession* session = nullptr;
+    opencdm_create_session(&source, keySystem.utf8().data(), openCDMLicenseType(licenseType), mimeType, reinterpret_cast<const uint8_t*>(m_initData->data()), m_initData->size(),
+        !customData->isEmpty() ? reinterpret_cast<const uint8_t*>(customData->data()) : nullptr, customData->size(), callbacksNotifier.callbacks(), &session);
+    if (!session)
         return;
+    m_session = ScopedSession(session, [](OpenCDMSession* session) {
+        opencdm_session_close(session);
+        opencdm_destruct_session(session);
+    });
+    m_id = String::fromUTF8(opencdm_session_id(m_session.get()));
+    callbacksNotifier.addObserver(m_session.get(), this);
+}
 
-    m_isValid = temporaryURLLength;
+CDMInstanceOpenCDM::Session::~Session()
+{
+    OCDMSessionCallbacksNotifier::singleton().removeObserver(this);
+}
 
+void CDMInstanceOpenCDM::Session::challengeGeneratedCallback(String&& message)
+{
     // We could do all operations with String but this way we can save some copies.
     size_t typePosition = message.find(":Type:");
```

Comment:
It seems that here is the first place where we have to parse the message that will come now as `Ref<SharedBuffer>`. We could copy to `String` before this but it is probably better try to use `g_strstr_len` to find the position.

Reply:
Ok. Will switch to `SharedBuffer` but I think I'll do the copying to `String` otherwise I'd have to change - apart from this place - lambdas in `loadSession()` and `removeSessionData()` and `sessionLoadFailureFromOpenCDM()` and `mediaKeyStatusFromOpenCDM()` and `parseResponseMessage()`. Also we'll have SharedBuffer -> String -> SharedBuffer switches since e.g. `parseResponseMessage()` takes `String` removes some stuff and returns `SharedBuffer`. This place does that too.